src/firstboot/firstboot.c

```diff
@@ -317,6 +318,166 @@ static int process_locale(void) {
+        if (r < 0)
+                return -errno;
+
+        return 0;
+}
+
+static int invoke_locale_gen(void) {
+        int r;
+
+        r = safe_fork("(locale-gen)", FORK_RESET_SIGNALS|FORK_RLIMIT_NOFILE_SAFE|FORK_DEATHSIG|FORK_LOG|FORK_WAIT, NULL);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                if (arg_root) {
+                        r = chdir(arg_root);
+                        if (r < 0) {
+                                log_error_errno(errno, "Failed to change working directory to %s: %m", arg_root);
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        r = chroot(".");
+                        if (r < 0) {
+                                log_error_errno(errno, "Failed to chroot to %s: %m", arg_root);
+                                _exit(EXIT_FAILURE);
+                        }
+                }
+
+                execlp("locale-gen", "locale-gen", NULL);
+
+                log_error_errno(errno, "Failed to execute locale-gen: %m");
+                _exit(EXIT_FAILURE);
+        }
+
+        return 0;
+}
+
+static int process_locale_gen(void) {
+        const char *etc_localegen;
+        int r;
+
+        r = find_binary("locale-gen", NULL);
+        if (r == -ENOENT) {
+                log_warning("locale-gen not found in PATH, skipping locale generation");
+                return 0;
+        }
+        if (r < 0)
+                return r;
+
+        etc_localegen = prefix_roota(arg_root, "/etc/locale.gen");
+        if (laccess(etc_localegen, F_OK) >= 0 && !arg_force)
+                return 0;
+
+        if (arg_copy_locale && arg_root) {
+
+                (void) mkdir_parents(etc_localegen, 0755);
+                r = copy_file("/etc/locale.gen", etc_localegen, 0, 0644, 0, 0, COPY_REFLINK);
+                if (r != -ENOENT) {
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to copy %s: %m", etc_localegen);
+
+                        log_info("%s copied.", etc_localegen);
+
+                        r = invoke_locale_gen();
+                        if (r < 0)
+                                return r;
+
+                        return 0;
+                }
+        }
+
+        /* No prompt here. We rely on the prompt done by process_locale_conf() instead. */
+
+        if (isempty(arg_locale))
+                return 0;
+
+        r = write_locale_gen(etc_localegen, arg_locale);
+        if (r < 0)
+                return log_error_errno(r, "Failed to write %s: %m", etc_localegen);
+
+        log_info("%s written", etc_localegen);
+
+        r = invoke_locale_gen();
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+static int process_locale(void) {
+        int r;
+
+        r = process_locale_conf();
+        if (r < 0)
+                return r;
+
+        r = process_locale_gen();
+        if (r < 0)
```

Comment:
There's an implicit ordering between process_locale_conf() and process_locale_gen() since process_locale_gen() depends on the prompt_locale() call made by process_locale_conf() so I thought to make that a little bit clearer by putting them in a helper function.

I'm not particularly convinced by that approach though so I can remove it if you think it doesn't add any value. A comment might provide the same utility.

Reply:
I removed the helper and added a comment instead.